### PR TITLE
adding contracts

### PR DESCRIPTION
**Pull request by julien51**
First commit!

### CODE REVIEW

File: contracts/UnlockPrimeHook.sol [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import "@unlock-protocol/contracts/dist/PublicLock/IPublicLockV13.sol";
+
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.27;
+
+// Uncomment this line to use console.log
+import "hardhat/console.sol";
```
Comment (clemsos) [inline contracts/UnlockPrimeHook.sol]: comment out？

File: contracts/UnlockPrimeHook.sol [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            uint refundAmount = refunds[from][0] + (valueInETH * 11) / 10; // 10% bonus
+            uint delay = block.timestamp + 60 * 60 * 24 * 14; // 2 weeks delay!
+
+            refunds[from] = [refundAmount, delay];
+            emit RefundSet(from, refundAmount, delay);
+        }
+    }
+
+    // Claim the refund!
+    function clainRefund() external {
```
Comment (clemsos) [inline contracts/UnlockPrimeHook.sol]:
```suggestion
    function claimRefund() external {
```

**Comment by julien51** [comment]
We still need to handle renewals here but I don't think we have a good hook for that right now.

**Review by clemsos — COMMENTED** [review]
lg so far. Looks like the claim is not tested though?

File: contracts/UnlockPrimeHook.sol [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            recordRefund(from, IPublicLockV13(msg.sender).keyPrice());
+        }
+    }
+
+    // This function is called by Unlock when a key is extended
+    function onKeyExtend(
+        uint tokenId,
+        address /* sender */,
+        uint /* newTimestamp */,
+        uint /* expirationTimestam p*/
```
Comment (clemsos) [inline contracts/UnlockPrimeHook.sol]:
```suggestion
        uint /* expirationTimestamp */
```

File: test/UnlockPrimeHook.ts [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+      await lock.extend(lockArgs.keyPrice, 1, deployer.getAddress(), "0x")
+    ).wait();
+
+    expect(await hook.refunds(await deployer.getAddress(), 0)).to.greaterThan(
+      refund
+    );
+    expect(await hook.refunds(await deployer.getAddress(), 1)).to.greaterThan(
+      delay
+    );
+  });
```
Comment (clemsos) [inline test/UnlockPrimeHook.ts:208]: Maybe add a test for claiming with renewal ?

**Review by clemsos — APPROVED** [review]
Lgtm ! Some minor comments but seems fine 👍

**Review by clemsos — APPROVED** [review]

**Pull request merged**